File: 04-Linked-List/Test/src/Array.java
public class Array<E> {

    private E[] data;

    private int size;

    public Array(int capacity){
        data = (E[])new Object[capacity];
        size = 0;
    }

    public Array(){
        this(10);
    }

    // 获取数组长度
    public int getCapacity(){
        return data.length;
    }

    // 获取已用长度
    public int getSize(){
        return size;
    }

    // 是否空数组
    public boolean isEmpty(){
        return size == 0;
    }

    // 在指定索引添加元素
    public void add(int index, E e){
        if (index < 0 || index > size)
            throw new IllegalArgumentException("index Illegal");

        if (size == getCapacity())
            resize(2*getCapacity());

        // index 之后的元素后移
        for (int i = size-1; i >= index; i--) {
            data[i+1] = data[i];
        }

        data[index] = e;
        size++;
    }

    public void addLast(E e){
        add(size,e);
    }

    public void addFirst(E e){
        add(0,e);
    }


    public E get(int index){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("index Illegal");

        return data[index];
    }

    public E getLast(){return get(size-1);}

    public E getFirst(){return get(0);}

    public void set(int index, E e){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("index Illegal");

        data[index] = e;
    }

    public boolean contains(E e){
        for (int i = 0; i < size; i ++){
            if (data[i].equals(e))
                return true;
        }
        return false;
    }

    // 返回元素e的索引值, 没找到返回 -1
    public int find(E e){
        for (int i = 0; i < size; i ++){
            if (data[i].equals(e))
                return i;
        }
        return -1;
    }

    public E remove(int index){
        if (index < 0 || index >= size)
            throw new IllegalArgumentException("index Illegal");

        E ret = data[index];
        for (int i = index+1; i < size; i++)
            data[i-1] = data[i];

        size--;
        if (size < getCapacity() / 4 && getCapacity()/2 != 0)
            resize(getCapacity() / 2);

        return ret;
    }

    public E removeFirst(){
        return remove(0);
    }

    public E removeLast(){
        return remove(size-1);
    }

    public void removeElement(E e){
        int index = find(e);
        if (index != -1)
            remove(index);
    }

    private void resize(int newCapacity){
        E[] newData = (E[])new Object[newCapacity];

        // 移动旧数据
        for (int i = 0; i < size; i++)
            newData[i] = data[i];

        data = newData;
    }
}

File: go/03-Stacks-and-Queues/02-Array-Stack/array/array.go
package array

import (
	"bytes"
	"fmt"
	"github.com/dingsw2019/play-with-data-struct/go/utils"
)

type Array struct {
	data []interface{}
	size int
}

func New(capacity int) *Array {
	return &Array{
		data: make([]interface{}, capacity),
	}
}

func (a *Array) GetCapacity() int {
	return len(a.data)
}

func (a *Array) GetSize() int {
	return a.size
}

func (a *Array) IsEmpty() bool {
	return a.size == 0
}

func (a *Array) Add(index int, e interface{}) {
	if index < 0 || index > a.size {
		panic("add failed, index out of range")
	}
	if a.size == len(a.data) {
		a.resize(2 * a.size)
	}
	for i := a.size - 1; i >= index; i-- {
		a.data[i+1] = a.data[i]
	}
	a.data[index] = e
	a.size++
}

func (a *Array) AddLast(e interface{}) {
	a.Add(a.size, e)
}

func (a *Array) AddFirst(e interface{}) {
	a.Add(0, e)
}

func (a *Array) Get(index int) interface{} {
	a.checkIndex(index)
	return a.data[index]
}

func (a *Array) GetLast() interface{} {
	return a.Get(a.size - 1)
}

func (a *Array) GetFirst() interface{} {
	return a.Get(0)
}

func (a *Array) Set(index int, e interface{}) {
	a.checkIndex(index)
	a.data[index] = e
}

func (a *Array) Contains(e interface{}) bool {
	for i := 0; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			return true
		}
	}
	return false
}

func (a *Array) Find(e interface{}) int {
	for i := 0; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			return i
		}
	}
	return -1
}

func (a *Array) FindAll(e interface{}) (indexes []int) {
	for i := 0; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			indexes = append(indexes, i)
		}
	}
	return
}

func (a *Array) Remove(index int) interface{} {
	a.checkIndex(index)
	e := a.data[index]
	for i := index + 1; i < a.size; i++ {
		a.data[i-1] = a.data[i]
	}
	a.size--
	a.data[a.size] = nil
	// 缩容, 避免左右震荡, capacity=1 时避免减成0
	if a.size == len(a.data)/4 && len(a.data)/2 != 0 {
		a.resize(len(a.data) / 2)
	}
	return e
}

func (a *Array) RemoveFirst() interface{} {
	return a.Remove(0)
}

func (a *Array) RemoveLast() interface{} {
	return a.Remove(a.size - 1)
}

func (a *Array) RemoveElement(e interface{}) {
	index := a.Find(e)
	if index != -1 {
		a.Remove(index)
	}
}

func (a *Array) RemoveAllElement(e interface{}) bool {
	start := a.Find(e)
	if start == -1 {
		return false
	}
	for i := start; i < a.size; i++ {
		if utils.Compare(a.data[i], e) == 0 {
			a.Remove(i)
		}
	}
	return true
}

func (a *Array) resize(newCapacity int) {
	newData := make([]interface{}, newCapacity)
	for i := 0; i < a.size; i++ {
		newData[i] = a.data[i]
	}
	a.data = newData
}

func (a *Array) checkIndex(index int) {
	if index < 0 || index >= a.size {
		panic(fmt.Sprintf("index out of range, [0,%d]", a.size))
	}
}

func (a *Array) String() string {
	var buf bytes.Buffer

	buf.WriteString(fmt.Sprintf("Array: size = %d, capacity = %d\n", a.size, len(a.data)))
	buf.WriteString("[")
	for i := 0; i < a.size; i++ {
		buf.WriteString(fmt.Sprintf("%v", a.data[i]))
		if i != a.size-1 {
			buf.WriteString(", ")
		}
	}
	buf.WriteString("]")
	return buf.String()
}
